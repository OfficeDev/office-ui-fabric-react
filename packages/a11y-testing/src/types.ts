export type PropValue = string | number | boolean | any;
export type Props = { [name: string]: PropValue };
export type AccessibilityBehavior = (props: Props) => any;

export interface Rule {
  stringify: () => string;
}

export interface Slot {
  name: string;
  props: Props[];
  expectAttribute: boolean;
  expectedAttribute: string;
  expectedValue: PropValue;
  description: string;
  hidden: boolean;
  afterEvent: string;
  afterEventData: any[];
  checkEvent: boolean;
}

export interface TestFacade {
  slotExists(slotName: string): boolean;
  attributeExists(slotName: string, attributeName: string): boolean;
  attributeHasValue(slotName: string, attributeName: string, value: PropValue): boolean;
  getAttributeValue(slotName: string, attribute: string): PropValue;
  forProps(props: Props): TestFacade;
  afterEvent(slotName: string, eventName: string, args: any[]): void;
}
