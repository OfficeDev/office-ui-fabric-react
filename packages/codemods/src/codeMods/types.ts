import { SourceFile } from 'ts-morph';

// tslint:disable-next-line: interface-name
export interface CodeModResult {
  success?: boolean;
}

// tslint:disable-next-line: interface-name
export interface Codemod<T = SourceFile> {
  /**
   * The version before which the codemod should run.
   * Must be valid semver.
   * Currently has no impact. Maybe remove?
   */
  version?: string;
  /**
   * A string to help identify the codemod.
   */
  name: string;
  /**
   * The actual function that should be run on any given file.
   * TODO, is there a possibility of codemods that would need to execute over mutiple files?
   */
  run: (file: T) => CodeModResult;
  /**
   * If not enabled, then this mod will not be conisdered to run. Only enable it once it's ready
   * to be applied in real world scenarios
   */
  enabled?: boolean;
}
