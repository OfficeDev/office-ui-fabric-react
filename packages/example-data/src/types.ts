export interface IFooProps {
  /** @deprecated */
  dep1?: string;
  /** @deprecated */
  dep2?: string;
  notDep?: string;
}
