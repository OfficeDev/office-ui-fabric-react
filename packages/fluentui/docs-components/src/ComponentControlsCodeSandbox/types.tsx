export type CodeSandboxImport = {
  module: any;
  version: string;
  required: boolean;
};

export type SourceManagerLanguage = 'js' | 'ts';
