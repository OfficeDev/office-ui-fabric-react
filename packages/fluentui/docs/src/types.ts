import { UseKnobOptions } from '@fluentui/docs-components';
import { ThemePrepared } from '@fluentui/react-northstar';
import { FluentComponentInfo, ComponentProp } from '@fluentui/react-docgen';

// !!!!!!!!!!
// Keep in sync with scripts/docs/types.ts!
// !!!!!!!!!!
export type ExampleSource = {
  js: string;
  ts: string;
};

/**
 * Component menu item shown in the site sidebar.
 * An array of menu items is saved in `componentMenu.json`.
 */
// !!!!!!!!!!
// Keep in sync with scripts/docs/types.ts!
// !!!!!!!!!!
export type ComponentMenuItem = {
  displayName: string;
  type: string;
};

export type KnobGeneratorOptions = {
  propName?: string;
  propDef: ComponentProp;
  componentInfo: FluentComponentInfo;
  theme: ThemePrepared;
};
export type KnobDefinition = UseKnobOptions<any> & { hook: Function };

export type KnobGenerator<T> = (options: KnobGeneratorOptions) => KnobDefinition;

export type KnobComponentGenerators<P> = Partial<Record<keyof P, KnobGenerator<any>>>;
