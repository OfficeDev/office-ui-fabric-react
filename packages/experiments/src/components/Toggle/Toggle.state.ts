import * as React from 'react';
import { IToggleProps, IToggleViewProps } from './Toggle.types';
import { BaseState } from '../../utilities/BaseState';

// Internal state will most likely include a subset of your ViewProps. This template just equates them to start with.
export type IToggleState = Pick<IToggleViewProps, 'checked' | 'onChange' | 'componentRef' | 'onClick'>;

export class ToggleState extends BaseState<IToggleProps, IToggleViewProps, IToggleState> {
  private _componentRef = React.createRef<HTMLElement>();

  constructor(props: ToggleState['props']) {
    super(props);

    this.state = {
      checked: !!props.defaultChecked,
      onChange: this._noop,
      componentRef: this._componentRef,
      onClick: this._onClick
    };
  }

  private _onClick = (ev: React.MouseEvent<HTMLElement>) => {
    const { disabled, checked: checkedProp, onChange } = this.props;
    const { checked } = this.state;

    if (!disabled) {
      // Only update the state if the user hasn't provided it.
      if (checkedProp === undefined) {
        this.setState({
          checked: !checked
        });
      }

      if (onChange) {
        onChange(ev, !checked);
      }
    }
  };

  private _noop(): void {
    /* no-op */
  }
}
