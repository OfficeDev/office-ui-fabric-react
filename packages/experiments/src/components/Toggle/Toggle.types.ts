import * as React from 'react';
import { IComponent, IStyleableComponentProps } from '../../Foundation';
import { IKeytipProps } from 'office-ui-fabric-react/lib/Keytip';
import { IStyle } from 'office-ui-fabric-react/lib/Styling';
import { IComponentAs } from '../../Utilities';

export type IToggleComponent = IComponent<IToggleProps, IToggleViewProps, IToggleStyles>;

/**
 * Toggle component props.
 */
export interface IToggleProps extends IStyleableComponentProps<IToggleViewProps, IToggleStyles> {
  /**
   * Render the root element as another type.
   */
  as?: IComponentAs<React.HTMLAttributes<HTMLElement>>;

  /**
   * A label for the toggle.
   */
  label?: string;

  /**
   * Text to display when toggle is ON.
   */
  onText?: string;

  /**
   * Text to display when toggle is OFF.
   */
  offText?: string;

  /**
   * Text for screen-reader to announce as the name of the toggle.
   */
  ariaLabel?: string;

  /**
   * Initial state of the toggle. If you want the toggle to maintain its own state, use this. Otherwise refer to `checked`.
   * @default true
   */
  defaultChecked?: boolean;

  /**
   * Checked state of the toggle. If you are maintaining state yourself, use this property. Otherwise refer to `defaultChecked`.
   * @default defaultChecked
   */
  checked?: boolean;

  /**
   * Optional disabled flag.
   */
  disabled?: boolean;

  /**
   * Callback issued when the value changes.
   */
  onChange?: (event: React.MouseEvent<HTMLElement>, checked?: boolean) => void;
}

export type IToggleViewProps = Pick<IToggleProps, 'disabled' | 'onChange'> &
  Required<Pick<IToggleProps, 'label' | 'onText' | 'offText' | 'ariaLabel' | 'checked'>> & {
    /**
     * Render the root element as another type.
     */
    as?: IComponentAs<React.HTMLAttributes<HTMLElement>>;

    /**
     * Optional callback to access the IToggle interface. Use this instead of ref for accessing
     * the public methods and properties of the component.
     */
    componentRef?: React.RefObject<HTMLElement>;

    /**
     * Toggle input callback triggered by mouse and keyboard input.
     */
    onClick?: (ev: React.MouseEvent<Element>) => void;

    /**
     * Optional keytip for this toggle
     */
    keytipProps?: IKeytipProps;

    /**
     * Root element class name.
     */
    className?: string;
  };

/**
 * Styles for the Toggle component.
 */
export interface IToggleStyles {
  /**
   * Root element.
   */
  root: IStyle;

  /**
   * Label element above the toggle.
   */
  label: IStyle;

  /**
   * Container for the toggle pill and the text next to it.
   */
  container: IStyle;

  /**
   * Pill, rendered as a button.
   */
  pill: IStyle;

  /**
   * Thumb inside of the pill.
   */
  thumb: IStyle;

  /**
   * Text next to the pill.
   */
  text: IStyle;
}
