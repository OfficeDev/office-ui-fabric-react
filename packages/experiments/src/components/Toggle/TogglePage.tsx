import * as React from 'react';
import { ExampleCard, IComponentDemoPageProps, ComponentPage, PropertiesTableSet } from '@uifabric/example-app-base';

import { ToggleExample } from './examples/Toggle.Example';
const ToggleExampleCode = require('!raw-loader!@uifabric/experiments/src/components/Toggle/examples/Toggle.Example.tsx') as string;

export class TogglePage extends React.Component<IComponentDemoPageProps, {}> {
  public render(): JSX.Element {
    return (
      <ComponentPage
        title=" Toggle"
        componentName=" Toggle"
        exampleCards={
          <div>
            <ExampleCard title=" Default Toggles" code={ToggleExampleCode}>
              <ToggleExample />
            </ExampleCard>
          </div>
        }
        propertiesTables={
          <PropertiesTableSet sources={[require<string>('!raw-loader!@uifabric/experiments/src/components/Toggle/Toggle.types.ts')]} />
        }
        overview={<div />}
        bestPractices={<div />}
        dos={
          <div>
            <ul>
              <li />
            </ul>
          </div>
        }
        donts={
          <div>
            <ul>
              <li />
            </ul>
          </div>
        }
        isHeaderVisible={this.props.isHeaderVisible}
      />
    );
  }
}
