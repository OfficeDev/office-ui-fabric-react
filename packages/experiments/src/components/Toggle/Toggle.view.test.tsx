import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { IToggleStyles } from './Toggle.types';
import { ToggleView } from './Toggle.view';
import { IProcessedStyleSet } from 'office-ui-fabric-react/lib/Styling';

// These tests will ensure that your styles regions have classname representation in snapshot output.
// (Also, classNames is a required prop for views, so we have to supply it for tests.)
const testToggleClassNames: IProcessedStyleSet<IToggleStyles> = {
  root: 'test-cn-root',
  label: 'test-cn-label',
  container: 'test-cn-container',
  pill: 'test-cn-pill',
  thumb: 'test-cn-thumb',
  text: 'test-cn-text',
  subComponentStyles: {}
};

// Views are just pure functions with no statefulness, which means they can get full code coverage
//    with snapshot tests exercising permutations of the props.
describe('ToggleView', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<ToggleView label="LabelProp" onText="On" offText="Off" checked={true} classNames={testToggleClassNames} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
