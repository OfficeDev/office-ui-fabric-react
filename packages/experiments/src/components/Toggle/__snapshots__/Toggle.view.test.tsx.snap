// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleView renders correctly 1`] = `
<div
  className="test-cn-root"
>
  <div>
    Label: 
    LabelProp
  </div>
</div>
`;
